### ISSUE TITLE
Support for Function::Parameters

### ISSUE THREAD

**Issue by RickMeasham** [Camelcade/Perl5-IDEA#1674]
In #1561, @pwr22 sort-of requested support for Function::Parameters. This ticket is to make it explicit. Please can we get support for Function::Parameters?

It provides a `method` alias to `sub` that allows additional arguments, and automates declaring `$self`. This, of course, should imply the `#@method` flag

It also provides a `fun` alias to `sub` that allows additional arguments, but doesn't declare `$self`

The module gets trickier than this so is probably not a simple piece of work. Anonymous `fun` and `method` are allowed. There are both positional parameters and named parameters. There is even the ability to change the names of `method` and `fun`. (Though seriously .. I wouldn't bother supporting that!)

**Comment by hurricup** [comment]
Same question: what is missing in current implementation of `method`/`func` (based on `Method::Signature::Simple`)

**Comment by RickMeasham** [comment]
I'll put together an example or two later in the day, but the first one that comes to mind is `fun` vs M:S:S's `func`. Akin to

```
    use Method::Signatures::Simple function_keyword => 'fun';
```

**Comment by RickMeasham** [comment]
![image](https://user-images.githubusercontent.com/79414/34231321-acd77cf0-e62f-11e7-8fd7-f2573e12c9e4.png)

In this snapshot we see that:

1. The `fun` keyword isn't recognised (per above, it's `func` in M:S:S)
1. * This leads to the function name being highlighted as a missing sub definition -- it thinks I'm _calling_ `test1`
1. * And the syntax error at the end of the line because it thinks we're calling a sub called `test1` and it wants a semicolon
2. The named parameters (denoted by a colon in front of them) are seen as syntax errors
3. The variables in the signature are highlighted as being undeclared

Here's the code to save retyping it if you want to test anything.

```perl
#!/usr/bin/env perl

use strict;
use warnings;
use 5.20.0;

use Function::Parameters;

fun test1 ( :$first, :$second ){
	say "Test 1";
	say "First: $first, Second: $second";
}
test1( first => 12, second => 15 );
test1( second => 15, first => 12 );


my $test2 = fun ( $first, $second ){
	say "Test 2";
	say "First: $first, Second: $second";
};
$test2->( 12, 15 );
```

**Comment by RickMeasham** [comment]
![image](https://user-images.githubusercontent.com/79414/34231564-aae4acbe-e630-11e7-8f22-abc94558b878.png)

Actually, there might be a bug in the Method::Signatures::Simple support. I'm not sure if it's related to some of the above issues or if I should create it as a separate issue.

You'll see that the two calls to `test1` are highlighted as calls to unknown subs

**Comment by hurricup** [comment]
Missed that there is a `fun` without `c`

**Comment by hurricup** [comment]
Few things:
1. Not so much reverce dependencies https://metacpan.org/requires/module/Function::Parameters?p=1&size=500&sort=%5B%5B2%2C1%5D%5D
2. We can't work with named arguments for now, there is no any architecture for this yet.

Tldr: not very soon.

**Comment by RickMeasham** [comment]
Thinking more about this (ignoring named parameters) I wonder if it would be possible to replace the current "func" and "method" handlers with a single more generic "sub-with-parameters" handler. Then you'd have a configuration item that let you list the sub-with-parameter names you use. Eg "func,fun,method,route".

My java isn't anywhere near strong enough to blindly do this, but it doesn't _look_ impossible.

**Comment by RickMeasham** [comment]
Actually, it looks like you may have already considered this, based on the comment here:

```bnf
/********************************************** Extensions for Method::Signatures *************************************/
// we can make this smareter and use keywords from settings or import opitions; We can't pin here because MooseX method works othewise
method_definition ::= 'method' sub_names_token [method_signature] [sub_attributes] parse_block
func_definition ::= 'func' sub_names_token [func_signature] [sub_attributes] parse_block
```

**Comment by hurricup** [comment]
`func` and `method` has slightly different syntax. But yes, still thinking about this.

**Comment by hurricup** [comment]
syntax for test 
```
    use overload
        fallback => 1,
        '""'     => method (@) { $self->{name} },
    ;
```

**Comment by ufobat** [comment]
Is it at least possible to turn off the errors that you see in the IDE, when you use function parameters?
For me I could accept that thereis no syntax check, as long as perl::critic is still working.

Can anyone help?

**Comment by hurricup** [comment]
@ufobat these are syntax errors, because plugin is not aware of such syntax. So - no, it's not possible :(

**Comment by hurricup** [comment]
Todo:
- [ ] fix rest of parser errors
- [ ]  add formatter tests
